File: day15.py

def go_one(n):
  lim = n 
  n = 0
  while n < lim: 
    yield c
    n+=1

def generator(seed,factor):
  tmin = 1<<31;
  prev = seed
  while True: 
    nxtv = (prev*factor)%(tmin-1) 
    yield nxtv 
    prev = nxtv

def generator3(seed,factor,mult):
  prev = seed
  while True: 
    g = generator(prev, factor)
    nxtv = next(g)
    while nxtv % mult != 0:
      nxtv = next(g)
    yield nxtv 
    prev = nxtv



def bitmatch(x,y,mask):
  return (x & mask) ^ (y & mask) == 0

if __name__ == "__main__": 
  SHOULD_RUN = True
  if SHOULD_RUN: 
    seed_a, seed_b = 679, 771 
    factor_a, factor_b = 16807, 48271 
    a,b = generator(seed_a, factor_a), generator(seed_b, factor_b)

    mask = 1<<15 
    mask = mask | mask >> 1
    mask = mask | mask >> 2
    mask = mask | mask >> 4
    mask = mask | mask >> 8

    c = 0
    for i in xrange(40000000):
      x,y = next(a), next(b)
      if bitmatch(x,y,mask):
        c += 1

    print 'part I:'
    print '  16-bit matches:', c

    c = 0
    aa,bb = generator3(seed_a, factor_a, 4), generator3(seed_b, factor_b, 8)
    for i in xrange(5000000):
      x,y  = next(aa), next(bb)
      if bitmatch(x,y, mask):
        c += 1

    print '\npart II:'
    print '  16-bit matches:', c
  else:
    print 'Change \'SHOULD_RUN\' to \'True\'.\nWARNING: Running time > 1 min.'

    
      

File: day15.c

#include <stdio.h>
#include <stdint.h>

#define LIM 40000000

int bitmatch(uint64_t, uint64_t, uint16_t);

int main() {
    unsigned int tmin = 1<<31;
    int c = 0;
    uint64_t prev_a = 679, prev_b = 771;
    int fa = 16807, fb = 48271;
    uint16_t m = 65535;

    for (int i=0; i<LIM; i++) {
        prev_a = (fa*prev_a)%(tmin-1);   
        prev_b = (fb*prev_b)%(tmin-1);   
        if (bitmatch(prev_a, prev_b, m)) c++;
    }

    printf("part I:\n");
    printf("  16-bit matches: %d\n", c);
    return 0;
}

int bitmatch(uint64_t x, uint64_t y, uint16_t m) {
    return ((x & m)^(y & m)) == 0;  
}

